cartography/intel/okta/oktaintel.py

```diff
@@ -0,0 +1,1225 @@
+    )
+
+
+def _sync_users_factors(neo4j_session, okta_org_id, okta_update_tag):
+    """
+    Sync user factors
+    :param neo4j_session: session with the Neo4j server
+    :param okta_org_id: okta organization id
+    :param okta_update_tag: The timestamp value to set our new Neo4j resources with
+    :return: Nothing
+    """
+
+    logger.debug("Syncing Okta User Factors")
+
+    factor_client = _create_factor_client(okta_org_id)
+    users = _get_user_id_from_graph(neo4j_session, okta_org_id)
+
+    for user_id in users:
+        user_factors = _get_factor_for_user_id(factor_client, user_id)
+        _ingest_user_factors(neo4j_session, user_id, user_factors, okta_update_tag)
+
+
+def _get_user_roles(api_client, user_id, okta_org_id):
+    """
+    Get user roles from Okta
+    :param api_client: api client
+    :param user_id: user to fetch roles from
+    :param okta_org_id: okta organization id
+    :return: Array of dictionary containing role properties
+    """
+
+    # https://developer.okta.com/docs/reference/api/roles/#list-roles
+    response = api_client.get_path(f'/{user_id}/roles')
+
+    return transform_user_roles_data(response.text, okta_org_id)
+
+
+def transform_user_roles_data(data, okta_org_id):
+    """
+    Transform user role data
+    :param data: data returned by Okta server
+    :param okta_org_id: okta organization id
+    :return: Array of dictionary containing role properties
+    """
+    role_data = json.loads(data)
+
+    user_roles = []
+
+    for role in role_data:
+        role_props = {}
+        role_props["label"] = role["label"]
+        role_props["type"] = role["type"]
+        role_props["id"] = "{}-{}".format(okta_org_id, role["type"])
+
+        user_roles.append(role_props)
+
+    return user_roles
+
+
+def _get_group_roles(api_client, group_id, okta_org_id):
+    """
+    Get user roles from Okta
+    :param api_client: api client
+    :param group_id: user to fetch roles from
+    :param okta_org_id: okta organization id
+    :return: Array of dictionary containing role properties
+    """
+
+    # https://developer.okta.com/docs/reference/api/roles/#list-roles-assigned-to-group
+    response = api_client.get_path(f'/{group_id}/roles')
+
+    return transform_group_roles_data(response.text, okta_org_id)
+
+
+def transform_group_roles_data(data, okta_org_id):
+    """
+    Transform user role data
+    :param data: data returned by Okta server
+    :param okta_org_id: okta organization id
+    :return: Array of dictionary containing role properties
+    """
+    role_data = json.loads(data)
+
+    user_roles = []
+
+    for role in role_data:
+        role_props = {}
+        role_props["label"] = role["label"]
+        role_props["type"] = role["type"]
+        role_props["id"] = "{}-{}".format(okta_org_id, role["type"])
+
+        user_roles.append(role_props)
+
+    return user_roles
+
+
+# TODO - Make this one fail gracefuly with warning as it requires super admin token
```

Comment:
Can you elaborate a bit on this?

Reply:
Addressed the todo already and removed it.

Basically, to get admin role you need full admin read/write token which most won't be able to get. That section of the intel module will fail gracefully with a warning vs error.